### DISCUSSION TITLE
Multipart/form-data post as client failed and MultipartEncoder is not fully tested!

### DISCUSSION THREAD

**Discussion by Rick0514** [wfrest/wfrest#111] [Q&A] [answered]
Http is totally new to me. Currently, I chose ```wfrest``` to realize some http function I want. Http server is fine but client seems to be not imperfect. Two problem remains:

#### 1. segment fault usnig MultipartEncoder.encode

see ```main/src/core/HttpContent.cc```:

```
std::string &MultiPartEncoder::encode()
{
    content_.clear();

    std::vector<std::pair<void *, WFFuture<ssize_t> *>> buf_fr_list;
    for(auto &file : files_) 
    {
        int fd = open(file.second.c_str(), O_RDONLY);
        if (fd >= 0)
        {
            size_t size = lseek(fd, 0, SEEK_END);
            void *buf = malloc(size);   
            WFFuture<ssize_t> fr = WFFacilities::async_pread(fd, buf, size, 0);
            buf_fr_list.push_back({buf, &fr});        
        } else 
        {
            buf_fr_list.push_back({nullptr, nullptr});
        }
    }

 for(int i = 0; i < files_.size(); i++)
    {

        ssize_t len = buf_fr_list[i].second->get();     // the get funtion cause segment fault!!!
...
```

I changed like below:

```
if (fd >= 0)
{
    size_t size = lseek(fd, 0, SEEK_END);
    void *buf = malloc(size);   
    WFFuture<ssize_t>* fr = new WFFuture<ssize_t>(WFFacilities::async_pread(fd, buf, size, 0));
    buf_fr_list.push_back({buf, fr});        
} else 
{
    buf_fr_list.push_back({nullptr, nullptr});
}
...
for(auto &buf_fr : buf_fr_list)
{
    free(buf_fr.first);
    if(buf_fr.second)  delete buf_fr.second;
}

```
and it works.

#### 2. Usnig the str encoded by MultipartEncoder.encode get empty string from server

here is my client code:

```
bool wait_flag = false;

client_task->set_callback([&](WFHttpTask *task)
{
    HttpRequest *req = task->get_req();
    HttpResponse *resp = task->get_resp();

    size_t body_len;
    const void* body;
    resp->get_parsed_body(&body, &body_len);

    HttpHeaderMap header(resp);
    std::string content_type = header.get("Content-Type");
    std::cout << "content type: " << content_type << std::endl;

    std::string resp_cont(static_cast<const char*>(body), body_len);
    // Json resp_cont_json = Json::parse(resp_cont);
    const char* errcode = resp->get_status_code();
    std::cout << "get store image link: " << resp_cont << std::endl;
    std::cout << "errcode: " << std::string(errcode) << std::endl;
    // std::cout << "info"
    // resp->get_reason_phrase();
    wait_flag = false;
});

client_task->get_req()->set_header_pair("Content-Type", "multipart/form-data; boundary=abcd");
// client_task->get_req()->set_header_pair("boundary", k_default_boundary);
// client_task->get_req()->set_header_pair("accept", "*/*");
client_task->get_req()->set_method("POST");

std::string filename = "/home/jtcx/mqtt-client/src/cpp/test/rev_http_file/aa.png";

MyMultiPartEncoder form;
form.add_file("test", filename);
std::string cont = form.encode();
client_task->get_req()->append_output_body_nocopy(cont.data(), cont.length()); 
wait_flag = true;
client_task->start();

while(wait_flag){   std::this_thread::sleep_for(std::chrono::milliseconds(100));    }

return 0;
```

however, i get empty string from the server.

I printed all the content that written to the body looks like below:

![image](https://user-images.githubusercontent.com/52786944/166610410-d3070e8e-80bd-4f75-bb11-a16cff284d85.png)

and got below: 

![image](https://user-images.githubusercontent.com/52786944/166610615-6b993d57-f3fa-42e1-89f4-f5697c24f3e8.png)

I also check the other library like ```libhv``` how it implement the multipart encoder, it seems no different.  I was wondering why it  happen??

**Comment by chanchann** [comment, marked as answer]
I'm very sorry for not removing this feature in time to cause some misunderstandings.
First of all, wfrest only focus on the server side.  In client side, you can use some lib like https://github.com/AndsonYe/MultipartEncoder. It's simple and easy, but it'a a sync way to wait for generating the content. If you have no performance requirements, you can use it.

The server side generally has higher performance requirementsin, so in order to achieve asynchronous effect, you can do like this:

```cpp
svr.GET("/form", [](const HttpReq *req, HttpResp *resp)
{
        MultiPartEncoder encoder;
        encoder.add_param("Filename", "1.jpg");
        encoder.add_file("test_1.txt", "./www/test_1.txt");
        encoder.add_file("test_2.txt", "./www/test_2.txt");
        resp->String(std::move(encoder));
});
```

**Reply by Rick0514** [reply]
Q2 is already solved, which is not a big deal. Thank you for the timely reply. At the first time I saw the project, I thought it maybe a  nice solution for both server and client side.

**Reply by chanchann** [reply]
This project is based on workflow.
I think workflow's client side is so clean that any changes are superfluous hhh.
I strongly recommend you to use workflow to write client side code.
